node/core/dispute-coordinator/src/initialized.rs

```diff
@@ -633,200 +630,149 @@ impl Initialized {
+				// Bounded because:
+				// 1. Only triggered twice per dispute.
+				// 2. Raising a dispute is costly (requires validation + recovery) by honest nodes,
+				// dishonest nodes are limited by spam slots.
+				// 3. Concluding a dispute is even more costly.
+				// Therefore it is reasonable to expect a simple vote request to succeed way faster
+				// than disputes are raised.
+				ctx.send_unbounded_message(
+					ApprovalVotingMessage::GetApprovalSignaturesForCandidate(candidate_hash, tx),
 				);
-
-				continue
+				match rx.await {
+					Err(_) => {
+						gum::warn!(
+							target: LOG_TARGET,
+							"Fetch for approval votes got cancelled, only expected during shutdown!"
+						);
+						intermediate_result
+					},
+					Ok(votes) => intermediate_result.import_approval_votes(&env, votes),
+				}
+			} else {
+				intermediate_result
 			}
+		};
+		let new_state = import_result.new_state();
 
-			match statement.statement() {
-				DisputeStatement::Valid(valid_kind) => {
-					let fresh = insert_into_statement_vec(
-						&mut votes.valid,
-						*valid_kind,
-						*val_index,
-						statement.validator_signature().clone(),
-					);
-
-					if !fresh {
-						continue
-					}
+		let is_included = self.scraper.is_candidate_included(&candidate_hash);
 
-					votes_changed = true;
-					self.metrics.on_valid_vote();
-				},
-				DisputeStatement::Invalid(invalid_kind) => {
-					let fresh = insert_into_statement_vec(
-						&mut votes.invalid,
-						*invalid_kind,
-						*val_index,
-						statement.validator_signature().clone(),
-					);
+		let potential_spam = !is_included && !new_state.is_confirmed() && !new_state.has_own_vote();
 
-					if !fresh {
-						continue
-					}
+		gum::trace!(
+			target: LOG_TARGET,
+			has_own_vote = ?new_state.has_own_vote(),
+			?potential_spam,
+			?is_included,
+			confirmed = ?new_state.is_confirmed(),
+			"Is spam?"
+		);
 
-					fresh_invalid_statement_issuers.push(*val_index);
-					votes_changed = true;
-					self.metrics.on_invalid_vote();
-				},
-			}
-		}
-
-		// Whether or not we know already that this is a good dispute:
-		//
-		// Note we can only know for sure whether we reached the `byzantine_threshold`  after
-		// updating candidate votes above, therefore the spam checking is afterwards:
-		let is_confirmed = is_included ||
-			was_confirmed ||
-			is_local || votes.voted_indices().len() >
-			byzantine_threshold(n_validators);
+		if !potential_spam {
+			// Former spammers have not been spammers after all:
+			self.spam_slots.clear(&(session, candidate_hash));
 
 		// Potential spam:
-		if !is_confirmed && !fresh_invalid_statement_issuers.is_empty() {
-			let mut free_spam_slots_available = true;
-			// Only allow import if all validators voting invalid, have not exceeded
-			// their spam slots:
-			for index in fresh_invalid_statement_issuers {
+		} else if !import_result.new_invalid_voters().is_empty() {
+			let mut free_spam_slots_available = false;
```

Comment:
The logic with `|=` and `&=` does not look very clear to express the idea.

Maybe it is better to have something like `let free_spam_slots_available = import_result.new_invalid_voters().iter().any(|index| self.spam_slots.add_unconfirmed(session, candidate_hash, *index))` ?

Reply:
If we decide to chage it, then it'll be `s/any/all/`

Reply:
I like.

Reply:
add_unconfirmed has side effects though and `any` is short circuiting. I would still like to increase counters on all validators.